knoten/sensor_utils.py

```diff
@@ -0,0 +1,296 @@
+from knoten import csm, utils, bundle
+from csmapi import csmapi
+import numpy as np
+
+def phase_angle(image_pt, sensor):
+    """
+    Computes and returns phase angle, in degrees for a given image point.
+   
+    Phase Angle: The angle between the vector from the intersection point to
+    the observer (usually the spacecraft) and the vector from the intersection
+    point to the illuminator (usually the sun).
+
+    Parameters
+    ----------
+    image_pt : tuple
+               Pair of x, y (sample, line) coordinates in pixel space
+
+    sensor : object
+             A CSM compliant sensor model object
+
+    Returns
+    -------
+     : np.ndarray
+            phase angle in degrees
+    """
+    if not isinstance(image_pt, csmapi.ImageCoord):
+        image_pt = csmapi.ImageCoord(*image_pt)
+
+    sensor_state = csm.get_state(sensor, image_pt)
+    ground_pt = csm.generate_ground_point(0.0, image_pt, sensor)
+  
+    sunEcefVec = sensor.getIlluminationDirection(ground_pt)
+    illum_pos = utils.Point(ground_pt.x - sunEcefVec.x, ground_pt.y - sunEcefVec.y, ground_pt.z - sunEcefVec.z)
+
+    vec_a = utils.Point(sensor_state["sensorPos"].x - ground_pt.x, 
+                        sensor_state["sensorPos"].y - ground_pt.y, 
+                        sensor_state["sensorPos"].z - ground_pt.z)
+
+    vec_b = utils.Point(illum_pos.x - ground_pt.x, 
+                        illum_pos.y - ground_pt.y, 
+                        illum_pos.z - ground_pt.z)
+  
+    return np.rad2deg(utils.sep_angle(vec_a, vec_b))
+
+def emission_angle(image_pt, sensor):
+    """
+    Computes and returns emission angle, in degrees, for a given image point.
+
+    Emission Angle: The angle between the surface normal vector at the
+    intersection point and the vector from the intersection point to the
+    observer (usually the spacecraft). The emission angle varies from 0 degrees
+    when the observer is viewing the sub-spacecraft point (nadir viewing) to 90
+    degrees when the intercept is tangent to the surface of the target body.
+    Thus, higher values of emission angle indicate more oblique viewing of the
+    target.
+
+    Parameters
+    ----------
+    image_pt : tuple
+               Pair of x, y (sample, line) coordinates in pixel space
+
+    sensor : object
+             A CSM compliant sensor model object
+
+    Returns
+    -------
+     : np.ndarray
+            emission angle in degrees
+    """
+    if not isinstance(image_pt, csmapi.ImageCoord):
+        image_pt = csmapi.ImageCoord(*image_pt)
+
+    sensor_state = csm.get_state(sensor, image_pt)
+    ground_pt = csm.generate_ground_point(0.0, image_pt, sensor)
+
+    normal = csm.get_surface_normal(sensor, ground_pt)
```

Comment:
This is also tied to the shapemodel rather than the sensor, providing some basic ellipsoid shape object to pass through to compute the normal rather than depending on the sensor to provide the necessary data. This is the case for any place where the call `get_surface_normal`

Reply:
I think this is an issue with me working with existing code in knoten. I originally needed to pass the sensor to call `get_radii` since it requires the sensor to obtain the ellipsoid. Are you saying that both these functions should instead pass in the ellipsoid object?

Reply:
I think there is a middle ground here that may be hard to explain over text, but I will try.

If we have a basic Ellipsoid shapemodel class, in many ways similar to the one ISIS has, we should be able to have a constructor that takes in the CSM sensor and populates the necessary fields to make the shapemodel.

It would likely be something like:
```
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c

    @classmethod
    def fromCSMSensor(CSMSensor):
        major, minor = get_radii(CSMSensor)
        return cls(major, major, minor)
```

Then the `Ellipsoid` class would expose a `intersect_surface` function and a `get_surface_normal` function. Something like:
```
my_shape.get_surface_normal(ground_pt)

and

my_shape.intersect_surface(sensor_state)
```

The class would more or less wrap the spice facilities that we currently have wrapped as functions. This also enables us to plan for more sophisticated shapemodels to eventually be added into the library with minimal friction (fingers crosssed)